libglvnd, ocl-icd, vulkan-loader: Add /run/opengl-driver(-32) to RUNPATH.

ambrop72 opened the pull request:
Previously we were relying on LD_LIBRARY_PATH to discover vendor driver libraries (libGL, ligGLX, libEGL, OpenCL and Vulkan). This has the problem that setuid programs (in particular VirtualBox) ignore LD_LIBRARY_PATH. Fix it by setting RUNPATH in various dispatch libraries to look in /run/opengl-driver(-32).

Setting this on the VirtualBoxVM binary (which would be simpler) would not solve the problem, the RUNPATH must be on the library which performs the dlopen(). The dlopen man page claims that the RUNPATH from the executable is used but that is incorrect, only RUNPATH from the module performing dlopen() is used.

Note that setting *RPATH* on the executable would work, but patchelf manipulates RUNPATH and not RPATH (RPATH is a legacy feature).

Fixes https://github.com/NixOS/nixpkgs/issues/22760
Fixes https://github.com/NixOS/nixpkgs/issues/18457.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
3D acceleration in VirtualBox does not work with NVidia drivers when hardening is enabled for VirtualBox (the default) (https://github.com/NixOS/nixpkgs/issues/22760)

###### Things done
Tested this on 19.03, with this change there is no error when starting a VM with 3D acceleration enabled and the VM log shows that OpenGL was initialized successfully.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

flokli commented:
does this mean we can drop https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/hardware/opengl.nix#L148-L150 too?

ambrop72 commented:
I think so, hope there isn't something else that uses it.

flokli commented:
@abbradar How does that sound to you?

ambrop72 commented:
Actually I think `LD_LIBRARY_PATH` may still be needed for libcuda, which does not have a dispatch library (AFAIK).

ambrop72 commented:
For CUDA I suppose we could apply the same RUNPATH correction to any code that dlopen's libcuda, but I am not at all familiar with CUDA things. Do apps normally dlopen libcuda directly, or are they linked to it, or is there a standard helper library that takes care of this?

ambrop72 commented:
Forgot about VDPAU though, where we have `libvdpau` as the dispatch library. From what I can tell, it does not need `LD_LIBRARY_PATH`, because we configure it with `--with-module-dir=${libGL_driver.driverLink}/lib/vdpau` and the code tries to `dlopen` `<module_dir>/lib/vdpau/libvdpau_<vendor>.sh`. So it should keep working if we get rid of `LD_LIBRARY_PATH`.

I will remove that `LD_LIBRARY_PATH` on my system and see if things break.

ambrop72 commented:
Added patching of some CUDA things.

With these patches, I have then removed that `LD_LIBRARY_PATH` settings from `opengl.nix` and do not see any breakage. KDE Plasma, Chromium, Blender, and Steam+games still work fine on my system with NVIDIA driver.

I suggest we apply the changes in this PR and also remove that `LD_LIBRARY_PATH`, then we can fix any breakages that people report by patching `RUNPATH` the same way. I would be really great to get rid of that `LD_LIBRARY_PATH` (it's a hack and it slows down program startup).

flokli commented:
v

flokli commented:
I gave this a test run on my system (19.03 with these two commits cherry-picked). Also removed adding to `LD_LIBRARY_PATH` from `opengl.nix` - no breakages observed - inclined to merge this :-)

@ambrop72, can you add the commit removing from `LD_LIBRARY_PATH` (plus some changelog entry) to this PR?

It should be safe to backport (only) 73f34cb138a317d3b3b144c381f8598d713484e5 to 19.03, too. cc @samueldr @lheckemann

ambrop72 commented:
Sorry forgot about this. Well now I know that maintainers can push to the PR branch by default :)
I agree to backport the first commit only,  to fix the VirtualBox problem.

abbradar commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -80,6 +81,16 @@ stdenv.mkDerivation rec {
         --prefix PYTHONPATH : ${pythonPackages.numpy}/${python.sitePackages}
     '';
 
+  # Set RUNPATH so that libcuda and libnvrtc in /run/opengl-driver(-32)/lib can be
```
Maybe this can be abstracted into a hook? The code is very similar. I think we could do with `patchelf`ing all stuff in `/bin` and `/lib` in it.

ambrop72 replied:
Is it really a good idea to add that RPATH entry to all binaries (once you include the hook)? I remember being bothered by some other hooks, like rpath stripping, not being configurable in that respect.

abbradar reviewed (commented):
Awesome idea! Regarding complete drop of `LD_LIBRARY_PATH`: I'm not sure now if that would break anything else; I'll try to remember what I encountered when I tried to remove it long time ago but don't count on this much. Bottom line is that there were _some_ problems but maybe they are not relevant anymore.

abbradar commented:
Nah, my idea is to make it non-propagating and only add it to libglvnd and friends. I just feel we'll have more derivations we'd need to fix so we want to factor out the code. 

On May 20, 2019 11:50:22 PM GMT+03:00, Ambroz Bizjak <notifications@github.com> wrote:
>ambrop72 commented on this pull request.
>
>
>
>> @@ -80,6 +81,16 @@ stdenv.mkDerivation rec {
>   --prefix PYTHONPATH : ${pythonPackages.numpy}/${python.sitePackages}
>     '';
> 
>+  # Set RUNPATH so that libcuda and libnvrtc in
>/run/opengl-driver(-32)/lib can be
>
>Is it really a good idea to add that RPATH entry to all binaries (once
>you include the hook)? I remember being bothered by some other hooks,
>like rpath stripping, not being configurable in that respect.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/60985#discussion_r285762966

-- 
Nikolay.

flokli commented:
@abbradar I'm fine with dropping the commit removing `LD_LIBRARY_PATH` for now, to not break too much at once :-)

Do you want to give the RUNPATH hook a try?

abbradar commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -80,6 +81,16 @@ stdenv.mkDerivation rec {
+    for program in $out/bin/blender $out/bin/.blender-wrapped; do
+      isELF "$program" || continue
+      origRpath=$(patchelf --print-rpath "$program")
+      patchelf --set-rpath "$origRpath:${libGL_driver.driverLink}/lib" "$program"
```
Maybe we want to put `driverLink` path first instead? This way if a package depends on Mesa or nvidia drivers directly it would still load system-wide drivers, not the ones it was linked with -- that's the behavior we arguably expect. It's especially important with nvidia because its driver libraries require exactly the same kernel module version and provide stable ABI.

ambrop72 replied:
There is only one libGL that can be linked to, that is the one from `libglvnd`. Neither the `mesa_drivers` nor `nvidia-x11` provide a libGL that can be linked to, only `libGL_vendor` that is dlopen-ed from `libglvnd`. I do not see any case where this RPATH adjustment would need to override some default library. Adding the RPATH at the end should reduce the performance impact to normal library loading.

abbradar commented on pkgs/development/libraries/vulkan-loader/default.nix:
```diff
@@ -17,7 +17,7 @@ stdenv.mkDerivation rec {
   };
 
-  nativeBuildInputs = [ pkgconfig ];
+  nativeBuildInputs = [ pkgconfig patchelf ];
```
`patchelf` is included in `stdenv` for Linux by default so no need to add it here.

abbradar commented:
@flokli I implemented a setup hook and added it everywhere. I only manually tested `libglvnd` for now.

abbradar commented:
@flokli Regarding `LD_LIBRARY_PATH`: I'm very much in favor of dropping it but have a gut feeling it would break something ;) I'll run a rebuild with this on my machine later. If it works okay too we could merge and fix breakages as they appear. @vcunat maybe you remember any problems w.r.t. replacing `LD_LIBRARY_PATH` with `RUNPATH`? We discussed `LD_LIBRARY_PATH` and OpenGL before.

abbradar commented:
I didn't mean libGL, rather any drivers themselves. Nvidia has this problem as I mentioned if one has nvidia_x11 in their RUNPATH for some reason, but also mesa_drivers. I think there were several cases where packages were linked to mesa_drivers or nvidia_x11 but I'll recheck this. 

On May 21, 2019 6:17:54 PM GMT+03:00, Ambroz Bizjak <notifications@github.com> wrote:
>ambrop72 commented on this pull request.
>
>
>
>> @@ -80,6 +81,16 @@ stdenv.mkDerivation rec {
>   --prefix PYTHONPATH : ${pythonPackages.numpy}/${python.sitePackages}
>     '';
> 
>+  # Set RUNPATH so that libcuda and libnvrtc in
>/run/opengl-driver(-32)/lib can be
>+  # found. See the explanation in libglvnd.
>+  postFixup = optionalString cudaSupport ''
>+    for program in $out/bin/blender $out/bin/.blender-wrapped; do
>+      isELF "$program" || continue
>+      origRpath=$(patchelf --print-rpath "$program")
>+      patchelf --set-rpath "$origRpath:${libGL_driver.driverLink}/lib"
>"$program"
>
>There is only one libGL that can be linked to, that is the one from
>`libglvnd`. Neither the `mesa_drivers` nor `nvidia-x11` provide a libGL
>that can be linked to, only `libGL_vendor` that is dlopen-ed from
>`libglvnd`. I do not see any case where this RPATH adjustment would
>need to override some default library. Adding the RPATH at the end
>should reduce the performance impact to normal library loading.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/60985#discussion_r286082442

-- 
Nikolay.

ambrop72 commented:
> I didn't mean libGL, rather any drivers themselves. Nvidia has this problem as I mentioned if one has nvidia_x11 in their RUNPATH for some reason, but also mesa_drivers. I think there were several cases where packages were linked to mesa_drivers or nvidia_x11 but I'll recheck this.

Well I didn't see any issue with appending RUNPATH. For libglvnd and other dispatch libraries the only thing that matters is that `dlopen` will find libraries like `libGL_vendor` and there is only going to be one package that has those, due to the name.

ambrop72 commented:
Yeah, I'll build and check a few things as well.

ambrop72 commented:
Tested that all packages with these R(UN)PATH adjustments build and get the correct R(UN)PATH (using `readelf -d`): libglvnd, ocl-icd, vulkan-loader, cudatoolkit, cudnn, nccl, blender(cudaSupport=true). Also cudatoolkit correctly gets RPATH instead of RUNPATH.

I will now rebase locally on top of channels/nixos-unstable and rebuild and test with my system (including the LD_LIBRARY_PATH removal).

ambrop72 commented:
While testing this I got hit by SDDM breakage in nixos-unstable (not caused by this, happens also on vanilla nixos-unstable, #61840). So I used LightDM instead, but with LightDM one still gets something in `LD_LIBRARY_PATH` (see [comment here](https://github.com/NixOS/nixpkgs/issues/11434#issuecomment-494673920)).

I will try removing that xserver `LD_LIBRARY_PATH` and fixing resulting breakage, but cannot do it for all drivers. But I don't think we should delay merging this (minus the `LD_LIBRARY_PATH` removal which can be done layer once more testing was done without it).

flokli commented:
@ambrop72 agreed. Could you squash things together a bit and drop 2f056afdc63ecd8595b43179aa6097aa6d53ece6?

abbradar committed:
addOpenGLRunpath: Add new hook for setting RUNPATH.

This hook allows to add NixOS driver libraries path to given ELF
objects' RUNPATH. We use it instead of settings RUNPATH manually
everywhere. It must be invoked in postFixup so that RUNPATH stripping
does not remove the path.

It puts the path first instead of last so that system-wide drivers
are always preferred.

ambrop72 committed:
libglvnd, ocl-icd, vulkan-loader: Add driver library paths to RUNPATH.

Previously we were relying on LD_LIBRARY_PATH to discover driver libraries (libGL, ligGLX, libEGL, OpenCL and Vulkan). This has the problem that setuid programs (in particular VirtualBox) ignore LD_LIBRARY_PATH. Fix it by setting RUNPATH in various dispatch libraries.

This is not needed for libvdpau because it is already configured to look for libraries in the driver paths.

Fixes https://github.com/NixOS/nixpkgs/issues/22760.

ambrop72 committed:
Add driver library path to some packages to find CUDA libraries.

This is to avoid relying on LD_LIBRARY_PATH for finding the CUDA driver libraries.

ambrop72 commented:
Rearranged commits, dropped the `LD_LIBRARY_PATH` removal. I think this is ready.

flokli commented:
Thanks!

matthewbauer commented on pkgs/build-support/add-opengl-runpath/setup-hook.sh:
```diff
@@ -0,0 +1,28 @@
+# actually sets RUNPATH not RPATH, which applies only to dependencies of the binary
+# it set on (including for dlopen), so the RUNPATH must indeed be set on these
+# libraries and would not work if set only on executables.
+addOpenGLRunpath() {
```
This shouldn't be a no-op on macOS. We can link to the OpenGL framework there I think.

flokli replied:
This should be a no-op, or we should link to the OpenGL framework?

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/software-rendering-using-mesa-not-possible/3023/8

arianvp commented on pkgs/development/libraries/vulkan-loader/default.nix:
```diff
@@ -17,17 +17,23 @@ stdenv.mkDerivation rec {
+  # Set RUNPATH so that driver libraries in /run/opengl-driver(-32)/lib can be found.
+  # See the explanation in addOpenGLRunpath.
+  postFixup = ''
+    addOpenGLRunpath $out/lib/libvulkan.so
```
This _should not_ be needed.

Vulkan Loader doesn't need the `openGLRunpath` hardcoded because it will look  up libraries not in the opengl path but in whatever   the ICD json points to.  So vulkan will never look in `/run/opengl-driver/lib`,
it will instead look in `/nix/store/y99l7729c0157a52y9qfzi3zg4f94b57-mesa-noglu-18.3.4-drivers/lib/libvulkan_intel.so`     because the `ICD` in `/run/opengl-driver/share/vulkan/icd.d` says so

```
cat /run/opengl-driver/share/vulkan/icd.d/intel_icd.x86_64.json 
{
    "ICD": {
        "api_version": "1.1.90",
        "library_path": "/nix/store/y99l7729c0157a52y9qfzi3zg4f94b57-mesa-noglu-18.3.4-drivers/lib/libvulkan_intel.so"
    },
    "file_format_version": "1.0.0"
}

```

ambrop72 replied:
It is currently needed for NVidia because the library_path is only a name:
```
$ cat /run/opengl-driver-32/share/vulkan/icd.d/nvidia.json 
{
    "file_format_version" : "1.0.0",
    "ICD": {
        "library_path": "libGLX_nvidia.so",
        "api_version" : "1.1.99"
    }
}
```
I think we should patch it and then remove the runpath.

flokli replied:
I agree. any other drivers that would need similar fixups? ;-)

ambrop72 replied:
Opened #62870 to set absolute paths in NVidia ICD files.
Also related:
- #62859: Fix that will enable actually dropping RUNPATH in dispatch libraries and LD_LIBRARY_PATH in `opengl.nix`.
- #62869: Will enable also dropping `XDG_DATA_DIRS` in `opengl.nix`.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/building-with-cuda-on-nixpkgs/4462/8

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/question-regarding-runpath-and-graphics-driver-configuration-files/75658/1
